README.md
# AdEspresso Monkeys [![Travis](https://img.shields.io/travis/adespresso/monkey-tools.svg?maxAge=2592000)](https://travis-ci.org/adespresso/monkey-tools/)
A collection of GreaseMonkey / TamperMonkey / Any monkey scripts we use in AdEspresso to simplify our daily lives

## Available monkeys
### PlanningPoker
* [Linkifies issue name](https://raw.githubusercontent.com/adespresso/monkey-tools/master/planningpoker/jira-issue.user.js)

### GitHub
* [Linkifies branch name](https://raw.githubusercontent.com/adespresso/monkey-tools/master/github/jira-issue.user.js)
* [JIRA sidebar](https://raw.githubusercontent.com/adespresso/monkey-tools/master/github/jira-sidebar-item.user.js)
* [JIRA reference fixer](https://raw.githubusercontent.com/adespresso/monkey-tools/master/github/jira-issue-reference.user.js)

### JIRA
* [GitHub sidebar](https://raw.githubusercontent.com/adespresso/monkey-tools/master/jira/github-sidebar-item.user.js)



CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.

## 2017-05-12
### Added
- Linkifies issue name for PlanningPoker leading to related JIRA issue
- Link Jira and GH tools to the website and plugin repositories as well

## 2017-03-23
### Added
- Remove noissue from PR title

## 2017-03-03
### Fixed
- GitHub scripts are now fully functional again after GH UI changes

## 2016-12-14
### Added
- Fills description for GitHub PRs related to JIRA issue
- Remove issue from PR title

## 2016-09-26
### Added
- eslint to enforce CS

## 2016-09-14
### Added
- JIRA sidebar with GitHub details

### Changed
- BUGS related branch name are now linkified too

## 2016-07-07
### Added
- GitHub sidebar with JIRA details

## 2016-07-01
### Added
- Linkifies branch name for GitHub leading to related JIRA issue



github/jira-issue.js
// ==UserScript==
// @name         GH Jira
// @namespace    http://adespresso.com/
// @version      1.1
// @description  Linkifies branch name
// @author       Massimiliano Cannarozzo
// @include      /^https://github.com/adespresso/adespresso/pull/\d+/
// @require      http://code.jquery.com/jquery-latest.js
// @require      https://gist.github.com/raw/2625891/waitForKeyElements.js
// ==/UserScript==
function linkifyBranchName() {
    var elem = $(".current-branch:contains('AAPP')").first();
    var text = elem.text();
    var issue = /^.+(AAPP-\d+)$/.exec(text)[1];
    var url = 'https://adespresso.atlassian.net/browse/'+issue;
    elem.attr('title', null);
    elem.html('<a href="'+url+'" target="_blank" title="See issue on Jira">'+elem.text()+'</a>');
}
(function() {
    waitForKeyElements(".current-branch:contains('AAPP')", linkifyBranchName);
})();



github/jira-issue.user.js
// ==UserScript==
// @name         GH Jira
// @namespace    http://adespresso.com/
// @version      1.5
// @description  Linkifies branch name
// @author       Massimiliano Cannarozzo
// @updateURL    https://raw.githubusercontent.com/adespresso/monkey-tools/master/github/jira-issue.user.js
// @include      /^https://github.com/adespresso/.+/pull/\d+/
// @require      http://code.jquery.com/jquery-latest.min.js
// @require      https://gist.github.com/raw/2625891/waitForKeyElements.js
// ==/UserScript==

function linkifyBranchName() {
    var elem = $(".head-ref:contains('AAPP'), .head-ref:contains('BUGS'), .head-ref:contains('AW'), .head-ref:contains('WWP')").last();
    var text = elem.text();
    var issue = /^.+((AAPP|BUGS|AW|WWP)-\d+)$/.exec(text)[1];
    var url = 'https://adespresso.atlassian.net/browse/' + issue;
    elem.attr('title', null);
    elem.html(
        '<img src="https://cloud.githubusercontent.com/assets/442835/16224961/b1bf3b26-37a4-11e6-96cd-7ebd6dad43fe.png" width="16px" style="vertical-align: middle; position: relative; top: -1px;">' +
        '<a href="' + url + '" target="_blank" title="See issue on Jira" style="vertical-align: middle">' + elem.text() + '</a>'
    );
}
(function () {
    waitForKeyElements(".head-ref:contains('AAPP'), .head-ref:contains('BUGS'), .head-ref:contains('AW'), .head-ref:contains('WWP')", linkifyBranchName);
})();


jira/github-sidebar-item.user.js
// ==UserScript==
// @name         Jira GH Issue Sidebar
// @namespace    http://adespresso.com/
// @version      1.0
// @description  Add GitHub sidebar item to Jira issue page
// @author       Massimiliano Cannarozzo
// @updateURL    https://raw.githubusercontent.com/adespresso/monkey-tools/master/jira/github-sidebar-item.user.js
// @grant        GM_xmlhttpRequest
// @include      /^https://adespresso.atlassian.net/browse/(AAPP|BUGS|AW|WWP)-\d+
// @require      https://code.jquery.com/jquery-latest.min.js
// @require      https://gist.github.com/raw/2625891/waitForKeyElements.js
// ==/UserScript==

var run = false;
var cachedResponse = null;
var prStatusClassMap = {
    MERGED: 'aui-lozenge-success',
    DECLINED: 'aui-lozenge-error',
    OPEN: 'aui-lozenge-complete',
};

function buildSidebarItem(id, children) {
    var sidebarItem = $(
        '<div id="' + id + '" class="module toggle-wrap">' +
        '<div id="greenhopper-agile-issue-web-panel_heading" class="mod-header">' +
        '<ul class="ops"></ul>' +
        '<h2 class="toggle-title">GitHub</h2>' +
        '</div>' +
        '<div class="mod-content"><ul class="item-details ghx-separated"></ul></div></div>'
    );

    sidebarItem.find('.item-details').append(children);
    return sidebarItem;
}

function injectOrUpdateGithubStatus(response) {
    response = response || cachedResponse;

    cachedResponse = response;
    if (response && !document.getElementById('jiraSidebarItem')) {
        var responseContent = JSON.parse(response.response).detail[0];
        var els = [];
        var prBranches = [];
        responseContent.pullRequests.forEach(pr => {
            prBranches.push(pr.source.branch);
            var el = $('<li><a href="' + pr.url + '">' +
                '<span class="aui-avatar-inner aui-icon aui-icon-small aui-iconfont-devtools-repository-forked"></span> ' +
                pr.id + '</a>&nbsp;<span class="aui-lozenge aui-lozenge-subtle ' + prStatusClassMap[pr.status] + '">' +
                pr.status + '</span></li>');
            els.push(el);
        });
        responseContent.branches.forEach(branch => {
            if (prBranches.indexOf(branch.name) >= 0) {
                return;
            }

            var el = $('<li><a href="' + branch.createPullRequestUrl + '/' + branch.name + '">' +
                '<span class="aui-avatar-inner aui-icon aui-icon-small aui-iconfont-devtools-repository"></span>' +
                '</a>&nbsp;' + branch.name + '</li>');
            els.push(el);
        });

        buildSidebarItem('jiraSidebarItem', els).appendTo($('#datesmodule'));
    }
    if (document.getElementById('viewissue-devstatus-panel')) {
        document.getElementById('viewissue-devstatus-panel').remove();
    }
}

function main() {
    // We don't want to make the XHR request more than once per page load.
    if (run) {
        injectOrUpdateGithubStatus();
        return;
    }
    run = true;

    var issue = $('#devstatus-container').attr('data-issue-id');
    var url = 'https://adespresso.atlassian.net/rest/dev-status/latest/issue/detail?issueId=' + issue + '&applicationType=github&dataType=pullrequest';
    GM_xmlhttpRequest({
        method: 'GET',
        url: url,
        onload: injectOrUpdateGithubStatus,
    });
}

(function () {
    waitForKeyElements('#devstatus-container', main);
})();



github/jira-issue-reference.user.js
// ==UserScript==
// @name         GH PR Jira Reference filler
// @namespace    http://adespresso.com/
// @version      1.2
// @description  Fills the reference to the JIRA issue automatically
// @author       Massimiliano Cannarozzo
// @updateURL    https://raw.githubusercontent.com/adespresso/monkey-tools/master/github/jira-issue-reference.user.js
// @include      /^https://github.com/adespresso/.+/compare/.+
// @require      https://code.jquery.com/jquery-latest.min.js
// @require      https://gist.github.com/raw/2625891/waitForKeyElements.js
// ==/UserScript==

function fillReference(issue) {
    $('#pull_request_body').val(`Ref. [${issue}](https://adespresso.atlassian.net/browse/${issue})`);
}
function removeReferenceFromTitle(issue) {
    var title = $('#pull_request_title')
        .val()
        .replace(issue.replace('-', ' ').toLowerCase(), '')
        .replace('noissue', '')
        .trim();
    $('#pull_request_title').val(title);
}
function fixReference() {
    var elem = $('.js-menu-target.branch > span').last();
    var text = elem.text();
    var matches = /^.*-([A-Z]+-\d+)$/.exec(text);
    if (matches.length > 0) {
        fillReference(matches[1]);
        removeReferenceFromTitle(matches[1]);
    }
}
(function () {
    waitForKeyElements('.js-menu-target', fixReference);
})();


github/notify-completed-checks.user.js
// ==UserScript==
// @name         Notify PR Checks ready
// @namespace    http://adespresso.com/
// @version      1.0
// @description  Shows notification of completed checks on a PR
// @author       Giorgio Cefaro
// @updateURL    https://raw.githubusercontent.com/adespresso/monkey-tools/master/github/notify-completed-checks.user.js
// @include      /^https://github.com/adespresso/adespresso/pull/\d+/
// @require      https://code.jquery.com/jquery-latest.min.js
// @require      https://gist.github.com/raw/2625891/waitForKeyElements.js
// @require      https://cdnjs.cloudflare.com/ajax/libs/notify.js/2.0.3/notify.min.js
// ==/UserScript==

var Notify = window.Notify.default;

function alertReady() {
    var notification = new Notify($('h1.gh-header-title .js-issue-title').text().trim(), {
        body: $('h4.status-heading').first().text().trim(),
        icon: $('.timeline-comment-avatar').first().attr('src'),
    });
    if (!Notify.needsPermission) {
        notification.show();
    } else if (Notify.isSupported()) {
        Notify.requestPermission(() => notification.show());
    }
}

(function () {
    if ($('.completeness-indicator-success').length === 0) {
        waitForKeyElements('.mergeability-details .completeness-indicator-success', alertReady);
    }
})();


github/jira-sidebar-item.user.js
// ==UserScript==
// @name         GH Jira Issue Sidebar
// @namespace    http://adespresso.com/
// @version      1.1
// @description  Add Jira sidebar item to GitHub pull request page
// @author       Massimiliano Cannarozzo
// @updateURL    https://raw.githubusercontent.com/adespresso/monkey-tools/master/github/jira-sidebar-item.user.js
// @grant        GM_getResourceText
// @grant        GM_xmlhttpRequest
// @include      /^https://github.com/adespresso/adespresso/pull/\d+/
// @require      https://code.jquery.com/jquery-latest.min.js
// @require      https://gist.github.com/raw/2625891/waitForKeyElements.js
// @resource     jira https://adespresso.atlassian.net/download/contextbatch/css/_super/batch.css?atlassian.aui.raphael.disabled=true&relative-url=true
// ==/UserScript==

var run = false;
var cachedResponse = null;
var jiraStyle = GM_getResourceText('jira');

function buildStyleElement() {
    var style = document.createElement('style');
    style.setAttribute('type', 'text/css');
    style.innerHTML = jiraStyle;

    return style;
}

function wrapSidebarItem(sidebarItem) {
    sidebarItem = $(sidebarItem);
    var content = sidebarItem.find('.content');

    var wrapper = {
        element: sidebarItem,
        prependTo: function (element) {
            return sidebarItem.prependTo(element);
        },
        append: function (stuff) {
            return content.append(stuff);
        },
        html: function (stuff) {
            return content.html(stuff);
        },
        find: function (selector) {
            return content.find(selector);
        },
    };

    return wrapper;
}

function getSidebarItem(id) {
    var sidebarItem = document.getElementById(id);

    if (sidebarItem) {
        return wrapSidebarItem(sidebarItem);
    } else {
        return null;
    }
}

function buildSidebarItem(id) {
    var sidebarItem = $(
        '<div id="' + id + '" class="discussion-sidebar-item">' +
        '   <h3 class="discussion-sidebar-heading">Jira</h3>' +
        '   <div class="content"></div>' +
        '</div>'
    );

    return wrapSidebarItem(sidebarItem);
}

function buildIssueLink(issue, summary) {
    var url = 'https://adespresso.atlassian.net/browse/' + issue;

    return $('<a href="' + url + '" target="_top" style="vertical-align: middle">' + issue + '</a>')
        .attr('title', summary);
}

function injectOrUpdateJiraStatus(response) {
    response = response || cachedResponse;
    cachedResponse = response;

    if (!response) {
        return;
    }

    var parent = document.getElementById('partial-discussion-sidebar');
    if (!parent) {
        return;
    }

    var responseContent = JSON.parse(response.response);
    var sidebarItem = getSidebarItem('jiraSidebarItem');
    if (!sidebarItem) {
        sidebarItem = buildSidebarItem('jiraSidebarItem');
        sidebarItem.prependTo(parent);
    }

    if (responseContent.errorMessages) {
        sidebarItem.html(responseContent.errorMessages.join('<br/>'));
    } else {
        var iframe = sidebarItem.find('iframe');
        if (!iframe.length) {
            iframe = $('<iframe width="200" height="20" scrolling="no" />')
                .css({
                    border: 'none',
                });
            sidebarItem.append(iframe);
        }
        var issue = responseContent.key;
        var summary = responseContent.fields.summary;
        var status = responseContent.fields.status;

        iframe = iframe.contents();
        iframe.find('head')
            .empty()
            .append(buildStyleElement());
        iframe.find('body')
            .empty()
            // We don't want Jira's gray default background.
            .css({ backgroundColor: 'transparent' })
            .append(buildIssueLink(issue, summary))
            .append(' <span class="aui-lozenge jira-issue-status-lozenge-' + status.statusCategory.colorName + '" style="vertical-align: middle">' + status.name + '</span>');
    }
}

function main() {
    var elem = $('.head-ref,h1').find(":contains('AAPP')").last();
    var text = elem.text();
    var issue = /(AAPP-\d+)/.exec(text)[1];

    // We don't want to make the XHR request more than once per page load/issue.
    if (run === issue) {
        injectOrUpdateJiraStatus();
        return;
    }
    run = issue;

    var url = 'https://adespresso.atlassian.net/rest/api/latest/issue/' + issue + '?fields=status,summary';
    GM_xmlhttpRequest({
        method: 'GET',
        url: url,
        onload: injectOrUpdateJiraStatus,
    });
}

(function () {
    waitForKeyElements(".head-ref:contains('AAPP'),h1:contains('AAPP')", main);
})();

